File: typescript_test_crate/tests/29_forward_enter_first.ts
class ForwardEnterFirstFrameEvent {
    public _message: string;
    public _parameters: Record<string, any> | null;
    public _return: any;

    constructor(message: string, parameters: Record<string, any> | null) {
        this._message = message;
        this._parameters = parameters;
        this._return = null;
    }
}


class ForwardEnterFirstCompartment {
    public state: string;
    public state_args: Record<string, any>;
    public state_vars: Record<string, any>;
    public enter_args: Record<string, any>;
    public exit_args: Record<string, any>;
    public forward_event: any;
    public parent_compartment: ForwardEnterFirstCompartment | null;

    constructor(state: string, parent_compartment: ForwardEnterFirstCompartment | null = null) {
        this.state = state;
        this.state_args = {  };
        this.state_vars = {  };
        this.enter_args = {  };
        this.exit_args = {  };
        this.forward_event = null;
        this.parent_compartment = parent_compartment;
    }

    public copy(): ForwardEnterFirstCompartment {
        const c = new ForwardEnterFirstCompartment(this.state, this.parent_compartment);
        c.state_args = {...this.state_args};
        c.state_vars = {...this.state_vars};
        c.enter_args = {...this.enter_args};
        c.exit_args = {...this.exit_args};
        c.forward_event = this.forward_event;
        return c;
    }
}


class ForwardEnterFirst {
    private _state_stack: Array<any>;
    private __compartment: ForwardEnterFirstCompartment;
    private __next_compartment: ForwardEnterFirstCompartment | null;
    private _return_value: any;
    private log: string[] =     [];

    constructor() {
        this._state_stack = [];
        this._return_value = null;
        this.log =         [];
        this.__compartment = new ForwardEnterFirstCompartment("Idle");
        this.__next_compartment = null;
        const __frame_event = new ForwardEnterFirstFrameEvent("$>", null);
        this.__kernel(__frame_event);
    }

    private __kernel(__e: ForwardEnterFirstFrameEvent) {
        // Route event to current state
        this.__router(__e);
        // Process any pending transition
        while (this.__next_compartment !== null) {
            const next_compartment = this.__next_compartment;
            this.__next_compartment = null;
            // Exit current state
            const exit_event = new ForwardEnterFirstFrameEvent("<$", this.__compartment.exit_args);
            this.__router(exit_event);
            // Switch to new compartment
            this.__compartment = next_compartment;
            // Enter new state (or forward event)
            if (next_compartment.forward_event === null) {
                const enter_event = new ForwardEnterFirstFrameEvent("$>", this.__compartment.enter_args);
                this.__router(enter_event);
            } else {
                // Forward event to new state
                const forward_event = next_compartment.forward_event;
                next_compartment.forward_event = null;
                if (forward_event._message === "$>") {
                    // Forwarding enter event - just send it
                    this.__router(forward_event);
                } else {
                    // Forwarding other event - send $> first, then forward
                    const enter_event = new ForwardEnterFirstFrameEvent("$>", this.__compartment.enter_args);
                    this.__router(enter_event);
                    this.__router(forward_event);
                }
            }
        }
    }

    private __router(__e: ForwardEnterFirstFrameEvent) {
        const state_name = this.__compartment.state;
        const handler_name = `_state_${state_name}`;
        const handler = (this as any)[handler_name];
        if (handler) {
            handler.call(this, __e);
        }
    }

    private __transition(next_compartment: ForwardEnterFirstCompartment) {
        this.__next_compartment = next_compartment;
    }

    public process() {
        const __e = new ForwardEnterFirstFrameEvent("process", null);
        this.__kernel(__e);
    }

    public get_counter(): number {
        this._return_value = null;
        const __e = new ForwardEnterFirstFrameEvent("get_counter", null);
        this.__kernel(__e);
        return this._return_value;
    }

    public get_log(): string[] {
        this._return_value = null;
        const __e = new ForwardEnterFirstFrameEvent("get_log", null);
        this.__kernel(__e);
        return this._return_value;
    }

    private _state_Working(__e: ForwardEnterFirstFrameEvent) {
        if (__e._message === "$>") {
            this.__compartment.state_vars["counter"] = 100;
            this.log.push("Working:enter")
        } else if (__e._message === "get_counter") {
            this._return_value = this.__compartment.state_vars["counter"];
            __e._return = this._return_value;
            return;
        } else if (__e._message === "get_log") {
            this._return_value = this.log;
            __e._return = this._return_value;
            return;
        } else if (__e._message === "process") {
            this.log.push("Working:process:counter=" + this.__compartment.state_vars["counter"].toString())
            this.__compartment.state_vars["counter"] = this.__compartment.state_vars["counter"] + 1
        }
    }

    private _state_Idle(__e: ForwardEnterFirstFrameEvent) {
        if (__e._message === "get_counter") {
            this._return_value = -1;
            __e._return = this._return_value;
            return;
        } else if (__e._message === "get_log") {
            this._return_value = this.log;
            __e._return = this._return_value;
            return;
        } else if (__e._message === "process") {
            const __compartment = new ForwardEnterFirstCompartment("Working", this.__compartment.copy());
            __compartment.forward_event = __e;
            this.__transition(__compartment);
            return;
        }
    }
}


function main(): void {
    console.log("=== Test 29: Forward Enter First ===");
    const s = new ForwardEnterFirst();

    if (s.get_counter() !== -1) {
        throw new Error("Expected -1 in Idle");
    }

    s.process();

    const counter = s.get_counter();
    const log = s.get_log();
    console.log(`Counter after forward: ${counter}`);
    console.log(`Log: ${JSON.stringify(log)}`);

    if (!log.includes("Working:enter")) {
        throw new Error(`Expected 'Working:enter' in log: ${log}`);
    }

    if (!log.includes("Working:process:counter=100")) {
        throw new Error(`Expected 'Working:process:counter=100' in log: ${log}`);
    }

    if (counter !== 101) {
        throw new Error(`Expected counter=101, got ${counter}`);
    }

    const enterIdx = log.indexOf("Working:enter");
    const processIdx = log.indexOf("Working:process:counter=100");
    if (enterIdx >= processIdx) {
        throw new Error(`$> should run before process: ${log}`);
    }

    console.log("PASS: Forward sends $> first for non-$> events");
}

main();

File: python_test_crate/tests/29_forward_enter_first.py
from typing import Any, Optional, List, Dict, Callable

class ForwardEnterFirstFrameEvent:
    def __init__(self, message: str, parameters):
        self._message = message
        self._parameters = parameters
        self._return = None


class ForwardEnterFirstCompartment:
    def __init__(self, state: str, parent_compartment = None):
        self.state = state
        self.state_args = {}
        self.state_vars = {}
        self.enter_args = {}
        self.exit_args = {}
        self.forward_event = None
        self.parent_compartment = parent_compartment

    def copy(self) -> 'ForwardEnterFirstCompartment':
        c = ForwardEnterFirstCompartment(self.state, self.parent_compartment)
        c.state_args = self.state_args.copy()
        c.state_vars = self.state_vars.copy()
        c.enter_args = self.enter_args.copy()
        c.exit_args = self.exit_args.copy()
        c.forward_event = self.forward_event
        return c


class ForwardEnterFirst:
    def __init__(self):
        self._state_stack = []
        self._return_value = None
        self.log =         []
        self.__compartment = ForwardEnterFirstCompartment("Idle")
        self.__next_compartment = None
        __frame_event = ForwardEnterFirstFrameEvent("$>", None)
        self.__kernel(__frame_event)

    def __kernel(self, __e):
        # Route event to current state
        self.__router(__e)
        # Process any pending transition
        while self.__next_compartment is not None:
            next_compartment = self.__next_compartment
            self.__next_compartment = None
            # Exit current state
            exit_event = ForwardEnterFirstFrameEvent("<$", self.__compartment.exit_args)
            self.__router(exit_event)
            # Switch to new compartment
            self.__compartment = next_compartment
            # Enter new state (or forward event)
            if next_compartment.forward_event is None:
                enter_event = ForwardEnterFirstFrameEvent("$>", self.__compartment.enter_args)
                self.__router(enter_event)
            else:
                # Forward event to new state
                forward_event = next_compartment.forward_event
                next_compartment.forward_event = None
                if forward_event._message == "$>":
                    # Forwarding enter event - just send it
                    self.__router(forward_event)
                else:
                    # Forwarding other event - send $> first, then forward
                    enter_event = ForwardEnterFirstFrameEvent("$>", self.__compartment.enter_args)
                    self.__router(enter_event)
                    self.__router(forward_event)

    def __router(self, __e):
        state_name = self.__compartment.state
        handler_name = f"_state_{state_name}"
        handler = getattr(self, handler_name, None)
        if handler:
            handler(__e)

    def __transition(self, next_compartment):
        self.__next_compartment = next_compartment

    def process(self):
        __e = ForwardEnterFirstFrameEvent("process", None)
        self.__kernel(__e)

    def get_counter(self) -> int:
        self._return_value = None
        __e = ForwardEnterFirstFrameEvent("get_counter", None)
        self.__kernel(__e)
        return self._return_value

    def get_log(self) -> list:
        self._return_value = None
        __e = ForwardEnterFirstFrameEvent("get_log", None)
        self.__kernel(__e)
        return self._return_value

    def _state_Working(self, __e):
        if __e._message == "$>":
            self.__compartment.state_vars["counter"] = 100
            self.log.append("Working:enter")
        elif __e._message == "get_counter":
            self._return_value = self.__compartment.state_vars["counter"]
            __e._return = self._return_value
            return
        elif __e._message == "get_log":
            self._return_value = self.log
            __e._return = self._return_value
            return
        elif __e._message == "process":
            self.log.append("Working:process:counter=" + str(self.__compartment.state_vars["counter"]))
            self.__compartment.state_vars["counter"] = self.__compartment.state_vars["counter"] + 1

    def _state_Idle(self, __e):
        if __e._message == "get_counter":
            self._return_value = -1
            __e._return = self._return_value
            return
        elif __e._message == "get_log":
            self._return_value = self.log
            __e._return = self._return_value
            return
        elif __e._message == "process":
            __compartment = ForwardEnterFirstCompartment("Working", parent_compartment=self.__compartment.copy())
            __compartment.forward_event = __e
            self.__transition(__compartment)
            return


def main():
    print("=== Test 29: Forward Enter First ===")
    s = ForwardEnterFirst()

    # Initial state is Idle
    assert s.get_counter() == -1, "Expected -1 in Idle"

    # Call process - should forward to Working
    # Correct behavior: $> runs first (inits counter=100, logs "Working:enter")
    # Then process runs (logs "Working:process:counter=100", increments to 101)
    s.process()

    # Check counter was initialized and incremented
    counter = s.get_counter()
    log = s.get_log()
    print(f"Counter after forward: {counter}")
    print(f"Log: {log}")

    # Verify $> ran first
    assert "Working:enter" in log, f"Expected 'Working:enter' in log: {log}"

    # Verify process ran after $>
    assert "Working:process:counter=100" in log, f"Expected 'Working:process:counter=100' in log: {log}"

    # Verify counter was incremented
    assert counter == 101, f"Expected counter=101, got {counter}"

    # Verify order: enter before process
    enter_idx = log.index("Working:enter")
    process_idx = log.index("Working:process:counter=100")
    assert enter_idx < process_idx, f"$> should run before process: {log}"

    print("PASS: Forward sends $> first for non-$> events")

if __name__ == '__main__':
    main()
